### ISSUE TITLE
[Feature Request] Is MSAL.Net compatible with workload identity federation credential?    

### ISSUE THREAD

**Issue by SCOMnewbie** [AzureAD/microsoft-authentication-library-for-dotnet#4368]
### MSAL client type

Confidential

### Problem Statement

Hi MSAL team,
After several hours in the docs, it seems MSAL.Net does not provide a way to interact with the workload identity federated credentials:

![image](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/assets/20821868/6842dbaf-144e-4fed-9807-2657f415147e)

![image](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/assets/20821868/10240cb9-71e7-48f7-860c-0613e508999d)

Is it me that missed the implementation? Or is it something you plan to provide later?

Cheers!

**Comment by bgavrilMS** [comment]
What is the other federation provider that you intend to use there? Is it external (e.g. GitHub) or internal (i.e. AAD - for multi-tenantcy scenarios)? 

I expect that can use `WithClientAssertion` instead of `WithSecret` or `WithCertificate` to pass in the token from the "other" identity provider.

**Comment by bgavrilMS** [comment]
We should update the `WithClientAssertion` docs to ensure that we mention this scenario, so that ~Goog..~ Bing indexes this. 

We should also have a code snippet on https://learn.microsoft.com/en-us/azure/active-directory/workload-identities/workload-identity-federation that showcases this.

**Comment by jmprieur** [comment]
There is also a specialization for managed identity federation with MSI (private preview) and managed identity federation for AKS. See https://github.com/AzureAD/microsoft-identity-web/wiki/v2.0#credentials-are-generalizing-certificates

**Comment by SCOMnewbie** [comment]
Thank you for the quick answer.  WithClientAssertion was my choice, but every time, I've received the error message like "your JWT is malformed", this is where I've started to think that it was not implemented. When you read https://learn.microsoft.com/en-us/entra/msal/dotnet/acquiring-tokens/web-apps-apis/confidential-client-assertions, the only example Microsoft provides is when you forge your token yourself, sign it and use it with WithClientAssertion which is not the same scenario.

Here what I try to do: https://azure.github.io/azure-workload-identity/docs/installation.html

Long story short, workload identities on a K8s cluster. I look forward for the example you can provide. I will take a look at
Microsoft.Identity.Web but it feels like a another dependency I don't want to have in my scenario.

**Comment by bgavrilMS** [comment]
Hi @SCOMnewbie 

Am I right in understanding that you are trying to setup a `workload running on any Kubernetes cluster (Azure Kubernetes Service (AKS), Amazon Web Services EKS, Google Kubernetes Engine (GKE), or on-premises). Establish a trust relationship between your user-assigned managed identity or app in Microsoft Entra ID and a Kubernetes workload`  as per https://learn.microsoft.com/en-us/azure/active-directory/workload-identities/workload-identity-federation#supported-scenarios ?

From an SDK perspective, this should be possible with both MSAL and Azure Identity (they mostly use MSAL underneath and provide higher level APIs). As you can see in the [examples](https://github.com/Azure/azure-workload-identity/tree/main/examples/msal-net/akvdotnet), the code just uses MSAL  via `WithClientAssertion(jwtFromAKS)` so I don't think there's anything wrong with your app code. 

The [docs](https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview?tabs=dotnet) do say `the AKS cluster acts as token issuer, Azure Active Directory uses OpenID Connect to discover public signing keys and verify the authenticity of the service account token before exchanging it for an Azure AD token.`. 

It is not clear from the docs, but you have to create an application in your AAD tenant and establish trust between it and your Kubernetes token issuer. Was this done?

<img width="711" alt="image" src="https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/assets/12273384/94df4a95-9c2e-4556-ab5b-9a56456cea75">

then...

<img width="470" alt="image" src="https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/assets/12273384/6823895b-7ea8-4119-af1f-a6730d7f19d4">


I think our friends from Azure SDK are in a better position to help, as they have documented this scenario end to end https://github.com/Azure/azure-workload-identity/tree/main

**Comment by SCOMnewbie** [comment]
Hi @bgavrilMS,
Thank you for the help, I've succeed to get my token. MSAL is definitely compatible with workload identity federated credentials!
Cheers

**Comment by pmaytak** [comment]
@bgavrilMS Can we reopen? I was thinking about adding a doc on MS Learn - this question have been asked a few times.

**Comment by bgavrilMS** [comment]
I also created a PR to update the Workload Identity sample docs: 

https://github.com/Azure/azure-workload-identity/pull/1146

**Comment by pmaytak** [comment]
The doc will be [here](https://learn.microsoft.com/entra/msal/dotnet/acquiring-tokens/web-apps-apis/workload-identity-federation) once this [PR](https://github.com/MicrosoftDocs/microsoft-authentication-library-dotnet/pull/261) is merged.
